### PR TITLE
Remove consumed from parse methods (per API review feedback)

### PR DESCRIPTION
**Pull request by JeremyKuhne**
- Move delegate creation to internal methods
- Add internal Advance for current span and tweak GetNextSpan
- Flip logic in binary read (now 15% faster)

### CODE REVIEW

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader.cs [hunk 1]
```diff
@@ -204,29 +204,26 @@ private void ResetReader()
...
+            SequencePosition previousNextPosition = _nextPosition;
+            while (_nextPosition.GetObject() != null && Sequence.TryGet(ref _nextPosition, out ReadOnlyMemory<T> memory, advance: true))
             {
-                SequencePosition previousNextPosition = _nextPosition;
-                while (Sequence.TryGet(ref _nextPosition, out ReadOnlyMemory<T> memory, advance: true))
+                _currentPosition = previousNextPosition;
+                if (memory.Length > 0)
+                {
+                    CurrentSpan = memory.Span;
+                    return;
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader.cs]: As a micro-optimization, consider avoiding a return within a loop and also having a single return point by using a goto.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs [hunk 2]
```diff
@@ -20,19 +20,20 @@ public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter, bool advancePastDel
         {
             ReadOnlySpan<T> remaining = UnreadSpan;
             int index = remaining.IndexOf(delimiter);
+
             if (index != -1)
             {
                 span = index == 0 ? default : remaining.Slice(0, index);
-                Advance(index + (advancePastDelimiter ? 1 : 0));
+                AdvanceCurrentSpan(index + (advancePastDelimiter ? 1 : 0));
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs]: Does the `count < 0` check really make much of a difference?

Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs]: Not huge, no. But the code is smaller.

**Comment by ahsonkhan** [comment]
> assert causes tests to silently skip in Test Explorer.

Yep. We get the following message in the test output:

> [10/12/2018 7:48:19 PM Error] The active test run was aborted. Reason: Assertion Failed

```text
...
at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Xunit.Sdk.TestInvoker`1.CallTestMethod(Object testClassInstance)
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<InvokeTestMethodAsync>b__1()
...
```

**Pull request merged**